<launch>
  <include file="$(find ego_planner)/launch/rviz.launch"/>
  
  <arg name="map_size_x" value="50.0"/>
  <arg name="map_size_y" value="50.0"/>
  <arg name="map_size_z" value="4.0"/>

  <arg name="drone_id" value="0"/>
  <arg name="obj_num" value="10" />

  <arg name="odom_topic" value="visual_slam/odom" />

  <!-- map -->
  <node pkg ="map_generator" name ="random_forest" type ="random_forest" output = "screen">
      <param name="map/x_size" value="36" />
      <param name="map/y_size" value="20" />
      <param name="map/z_size" value="5" />
      <param name="map/resolution" value="0.1"/>
      <param name="ObstacleShape/seed" value="1"/>
      <param name="map/obs_num" value="200"/>
      <param name="ObstacleShape/lower_rad" value="0.5"/>
      <param name="ObstacleShape/upper_rad" value="0.7"/>
      <param name="ObstacleShape/lower_hei" value="0.0"/>
      <param name="ObstacleShape/upper_hei" value="3.0"/>
      <param name="map/circle_num" value="200"/>
      <param name="ObstacleShape/radius_l" value="0.7"/>
      <param name="ObstacleShape/radius_h" value="0.5"/>
      <param name="ObstacleShape/z_l" value="0.7"/>
      <param name="ObstacleShape/z_h" value="0.8"/>
      <param name="ObstacleShape/theta" value="0.5"/>
      <param name="pub_rate" value="1.0"/>
      <param name="min_distance" value="0.8"/>
  </node>

    <!-- main algorithm params -->
  <include file="$(find ego_planner)/launch/advanced_param.xml">

    <arg name="drone_id" value="$(arg drone_id)"/>

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="odometry_topic" value="$(arg odom_topic)"/>

    <arg name="obj_num_set" value="$(arg obj_num)" />

    <!-- camera pose: transform of camera frame in the world frame -->
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- don't set cloud_topic if you already set these ones! -->
    <arg name="camera_pose_topic" value="pcl_render_node/camera_pose"/>
    <arg name="depth_topic" value="pcl_render_node/depth"/>

    <!-- topic of point cloud measurement, such as from LIDAR  -->
    <!-- don't set camera pose and depth, if you already set this one! -->
    <arg name="cloud_topic" value="pcl_render_node/cloud"/>

    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="321.04638671875"/>
    <arg name="cy" value="243.44969177246094"/>
    <arg name="fx" value="387.229248046875"/>
    <arg name="fy" value="387.229248046875"/>

    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="2.0" />
    <arg name="max_acc" value="3.0" />

    <!--always set to 1.5 times grater than sensing horizen-->
    <arg name="planning_horizon" value="4" /> 

    <arg name="use_distinctive_trajs" value="false" />

    <!-- 1: use 2D Nav Goal to select goal  -->
    <!-- 2: use global waypoints below  -->
    <arg name="flight_type" value="1" />
    
    <!-- global waypoints -->
    <!-- It generates a piecewise min-snap traj passing all waypoints -->
    <arg name="point_num" value="1" />

    <arg name="point0_x" value="20" />
    <arg name="point0_y" value="0" />
    <arg name="point0_z" value="2.0" />

    <arg name="point1_x" value="-20" />
    <arg name="point1_y" value="5" />
    <arg name="point1_z" value="1.0" />

    <arg name="point2_x" value="15.0" />
    <arg name="point2_y" value="0.0" />
    <arg name="point2_z" value="1.0" />

    <arg name="point3_x" value="0.0" />
    <arg name="point3_y" value="-15.0" />
    <arg name="point3_z" value="1.0" />

    <arg name="point4_x" value="-15.0" />
    <arg name="point4_y" value="0.0" />
    <arg name="point4_z" value="1.0" />
    
  </include>

  <!-- trajectory server -->
  <node pkg="ego_planner" name="drone_$(arg drone_id)_traj_server2" type="traj_server2" output="screen">
      <remap from="/planning/track_traj" to="drone_$(arg drone_id)_planning/track_traj"/>
      <remap from="~planning/bspline" to="drone_$(arg drone_id)_planning/bspline"/>
      <param name="traj_server/time_forward" value="1.0" type="double"/>
  </node>

  <!-- use simulator -->
  <include file="$(find ego_planner)/launch/simulator2.xml">
      <arg name="drone_id" value="$(arg drone_id)"/>
      <arg name="map_size_x_" value="$(arg map_size_x)"/>
      <arg name="map_size_y_" value="$(arg map_size_y)"/>
      <arg name="map_size_z_" value="$(arg map_size_z)"/>
      <arg name="init_x_" value="-24"/>
      <arg name="init_y_" value="0"/>
      <arg name="init_z_" value="0.1"/>
      <arg name="odometry_topic" value="$(arg odom_topic)" />
  </include>

  <node pkg="px4_ctrl" name="px4_mpc" type="px4_mpc_node" args="$(find px4_ctrl)/src/mpc_T05_dt01.so" output="screen">
    <remap from="/planning/track_traj" to="drone_$(arg drone_id)_planning/track_traj"/>
    <remap from="/px4/control_u" to="drone_$(arg drone_id)_px4/control_u"/>
    <remap from="/px4/state" to="drone_$(arg drone_id)_$(arg odom_topic)"/>
  </node>

</launch>